### PR TITLE
Gnome: Add appstream refresh so new apps are picked up

### PR DESCRIPTION
**Pull request by cidkidnix**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bjornfor** [comment]
So both this PR and https://github.com/NixOS/nixpkgs/pull/118100 solves the same issue (but differently)?

**Comment by cidkidnix** [comment]
I believe yes, but I'm also somewhat sure this is how gnome expects it to be done normally (eg. Arch, Fedora, and Ubuntu update appstream metadata on application install).

File: nixos/modules/services/x11/desktop-managers/gnome.nix [hunk 1]
```diff
@@ -314,8 +314,13 @@ in
       # Override GSettings schemas
       environment.sessionVariables.NIX_GSETTINGS_OVERRIDES_DIR = "${nixos-gsettings-desktop-schemas}/share/gsettings-schemas/nixos-gsettings-overrides/glib-2.0/schemas";
 
-       # If gnome is installed, build vim for gtk3 too.
+      # If gnome is installed, build vim for gtk3 too.
       nixpkgs.config.vim.gui = "gtk3";
+
```
Comment (SuperSandro2000) [inline nixos/modules/services/x11/desktop-managers/gnome.nix:319]:
```suggestion

```

**Review by SuperSandro2000 — COMMENTED** [review]
I think this solution is easier to maintain than patching glib.

**Comment by cidkidnix** [comment]
The only issue is nix-env installing packages won't be picked up (unless it's possible to add a hook to the end of nix-env installs)

**Comment by bjornfor** [comment]
Let's squash commits and merge?

**Comment by cidkidnix** [comment]
I'd still add this as most package managers do this already, but it's not quite ass effective as I first thought, oddly only some packages are affected by it

**Comment by SuperSandro2000** [comment]
Please target master and squash your commits.

**Comment by jtojnar** [comment]
I was not able to reproduce this on `nixos-unstable` (GNOME 40). I tried adding `qalculate-gtk` to `environment.systemPackages` and the code from this PR to my NixOS configuration but Shell still did not list qalculate (the package contains both desktop and appstream file).

I also tried to trigger GNOME Shell reload locally but it does not seem to do anything with the cache eithert:

```ShellSession
$ nix shell -f . appstream -c sudo --preserve-env appstreamcli refresh-cache --force
AppStream cache update completed successfully.
$ nix shell -f . appstream -c appstreamcli status
AppStream Status:
Version: 0.14.4

Distribution metadata:
 /usr/share/app-info
  - Empty.

 /var/lib/app-info
  - Empty.

 /var/cache/app-info
  - Empty.

Metainfo files:
  - Empty.

Summary:
We have information on 0 software components.
```

**Comment by cidkidnix** [comment]
It refreshes some apps but not all of them, we probably should close in favor of https://github.com/NixOS/nixpkgs/pull/118100

**Comment by cidkidnix** [comment]
I'm closing on the comment I have said above, feel free to reopen in a different PR with the changes mentioned.

**Pull request closed without merging**